### PR TITLE
#46_validate_round_scores

### PR DESCRIPTION
**Pull request by MariusTulbure**
Fixes https://github.com/robert-damoc/game-stats/issues/46
Added validation for scores on every round_type.
The first 2 validations required were made in the last 2 task.

### CODE REVIEW

File: app/models/round.rb [hunk 1]
```diff
@@ -42,4 +48,70 @@ def update_positions
...
+      validate_round_type_score(NEGATIVE_VALUE, (0..NEGATIVE_VALUE).step(-400).to_a, -400)
+    when 'ten'
+      validate_round_type_score(POSITIVE_VALUE, (0..POSITIVE_VALUE).step(400).to_a, 400)
+    when 'queens'
+      validate_round_type_score(NEGATIVE_VALUE, (0..NEGATIVE_VALUE).step(-100).to_a, 100)
+    when 'diamonds'
+      expected_value = NEGATIVE_DIAMONDS * game.game_players.count * 2
+      validate_round_type_score(expected_value, (0..NEGATIVE_VALUE).step(-50).to_a, -50)
+    when 'totale_minus', 'totale_plus'
+      validate_totale_scores(round_type)
```
Comment (robert-damoc) [inline app/models/round.rb:64]: `round_type` is an attribute on the current instance. Why do you need to pass it around functions?

File: app/models/round.rb [hunk 2]
```diff
@@ -42,4 +48,70 @@ def update_positions
...
+    when 'rentz_minus', 'rentz_plus'
+      validate_rentz_scores(round_type)
+    end
```
Comment (robert-damoc) [inline app/models/round.rb:67]: This looks too complicated, and too hard to follow. Lots of un-named parameters, the constants look a bit weird/unclear name

Comment (MariusTulbure) [reply app/models/round.rb]: This approach is cleaner and/or easier to understand?

File: app/models/round.rb [hunk 3]
```diff
@@ -42,4 +47,55 @@ def update_positions
...
+  def validate_scores
+    case round_type
+    when 'ten' then validate_round_type_score(STANDARD_VALUE, STANDARD_VALUE)
+    when 'king' then validate_round_type_score(-STANDARD_VALUE, -STANDARD_VALUE)
+    when 'queens' then validate_round_type_score(-STANDARD_VALUE, -QUEENS_VALUE)
+    when 'totale_minus' then validate_totale_scores(-STANDARD_VALUE, -DIAMONDS_VALUE)
+    when 'totale_plus' then validate_totale_scores(STANDARD_VALUE, DIAMONDS_VALUE)
+    when 'rentz_minus' then validate_rentz_scores(-STANDARD_VALUE)
+    when 'rentz_plus'  then validate_rentz_scores(STANDARD_VALUE)
+    when 'diamonds' then validate_diamonds_scores(-DIAMONDS_VALUE)
```
Comment (robert-damoc) [inline app/models/round.rb:60]: I think you should used named arguments

File: app/models/round.rb [hunk 4]
```diff
@@ -42,4 +47,55 @@ def update_positions
...
+    validate_round_type_score(expected_value, step_value)
+  end
+
+  def validate_totale_scores(standard_value, step_value)
+    expected_value = (standard_value * 3) + (step_value * game.game_players.count * 2)
+    validate_round_type_score(expected_value, step_value)
+  end
+
+  def validate_rentz_scores(step_value)
+    expected_value = step_value * game.game_players.count * (game.game_players.count - 1) / 2
```
Comment (robert-damoc) [inline app/models/round.rb:75]: I think the formula is `N * (N + 1) / 2`

Comment (robert-damoc) [reply app/models/round.rb:75]: Oh yeah, because lowest is 0, not 400 :+1:

**Review by robert-damoc — APPROVED** [review]
